### PR TITLE
[SPARK-13340][ML] PolynomialExpansion, Normalizer and ElementwiseProduct should validate input type

### PR DESCRIPTION
**Pull request by grzegorz-chilkiewicz**

### CODE REVIEW

**Comment by MLnick** [comment]
Jenkins test this please

File: mllib/src/test/scala/org/apache/spark/ml/feature/NormalizerSuite.scala [hunk 1]
```diff
@@ -105,6 +105,23 @@ class NormalizerSuite extends SparkFunSuite with MLlibTestSparkContext with Defa
...
   }
 
+  test("Normalization should throw adequate exception on input type mismatch") {
+    val data = Seq(Tuple1("string value"))
+
+    val df = sqlContext.createDataFrame(data).toDF("features")
+
+    val normalizer = new Normalizer()
+      .setInputCol("features")
+      .setOutputCol("polyFeatures")
```
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/feature/NormalizerSuite.scala]: very minor nit - but `polyFeatures` as a col name should probably be `normFeatures` or something like that

Comment (grzegorz-chilkiewicz) [inline mllib/src/test/scala/org/apache/spark/ml/feature/NormalizerSuite.scala]: Fixed: `normalized_features` is the column name used in other tests in NormalizerSuite.scala,
thanks!

**Comment by MLnick** [comment]
ok to test

**Comment by MLnick** [comment]
Thanks @grzegorz-chilkiewicz. I see `ElementwiseProduct` also needs `validateInputType`.

I'm wondering if it makes sense to do the default validation in the abstract class?

Something like the following in `UnaryTransformer`:
```scala
  /**
   * Validates the input type. Throw an exception if it is invalid.
   */
  protected def validateInputType(inputType: DataType): Unit = {
    val inType = classOf[IN].getSimpleName
    require(inputType.isInstanceOf[IN], s"Input type must be $inType but got $inputType.")
  }
```

@jkbradley @mengxr what do you think about this?

**Comment by jkbradley** [comment]
@MLnick We might be able to provide a standard implementation, but we would need to use Catalyst's ScalaReflection methods since ```inputType``` is a Catalyst DataType but ```classOf[IN]``` is a Scala class.  We also might have to override it in a few cases, such as transformers which should accept all NumericTypes.

**Comment by MLnick** [comment]
@jkbradley ah you're right. I was just thinking it would avoid little issues such as these creeping in.

@grzegorz-chilkiewicz if you'd like to look into a standard impl, let me know. Otherwise, if you can also do this for `ElementwiseProduct` it looks fine to me to merge.

**Comment by grzegorz-chilkiewicz** [comment]
I've broken here something during making change - please do not review this

File: mllib/src/main/scala/org/apache/spark/ml/feature/ElementwiseProduct.scala [hunk 2]
```diff
@@ -55,6 +55,11 @@ class ElementwiseProduct(override val uid: String)
     elemScaler.transform
   }
 
+  override protected def validateInputType(inputType: DataType): Unit = {
+    super.validateInputType(inputType)
+    require(inputType.isInstanceOf[VectorUDT], s"Input type must be VectorUDT but got $inputType.")
```
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/feature/ElementwiseProduct.scala]: This same pattern occurs in multiple places now. We have `SchemaUtils.checkColumnType`, so I think we should create a new method `SchemaUtils.checkDataType` in a similar manner, to harmonize the functionality and in particular the default error message across all present and future usages.

**Comment by MLnick** [comment]
retest this please

**Pull request closed without merging**